common/cuda_hip/reorder/rcm_kernels.hpp.inc

```diff
@@ -0,0 +1,614 @@
+        const auto col = col_idxs[nz];
+        const auto neighbor_level = node_levels[col];
+        if (neighbor_level < cur_level) {
+            min_neighbor = min(min_neighbor, inv_permutation[col]);
+        }
+    }
+    min_neighbors[target] = min_neighbor;
+}
+
+
+template <typename IndexType>
+__global__ __launch_bounds__(default_block_size) void build_permutation_level(
+    const IndexType level_begin, const IndexType level_end,
+    const IndexType* const level_nodes, IndexType* const inv_permutation)
+{
+    const auto target = thread::get_thread_id_flat() + level_begin;
+    if (target >= level_end) {
+        return;
+    }
+    inv_permutation[level_nodes[target]] = target;
+}
+
+
+template <typename IndexType>
+void sort_levels(std::shared_ptr<const DefaultExecutor> exec,
+                 const IndexType num_rows, const IndexType* const row_ptrs,
+                 const IndexType* const col_idxs,
+                 const IndexType* const degrees,
+                 const components_data<IndexType>& comps,
+                 ubfs_levels<IndexType>& levels, IndexType* const permutation)
+{
+    const auto policy = thrust_policy(exec);
+    array<IndexType> inv_permutation_array{exec,
+                                           static_cast<size_type>(num_rows)};
+    array<IndexType> key_array{exec, static_cast<size_type>(num_rows)};
+    levels.ptrs.set_executor(exec->get_master());
+    const auto num_levels = levels.num_levels;
+    const auto level_ptrs = levels.ptrs.get_data();
+    const auto level_nodes = levels.nodes.get_data();
+    const auto node_levels = levels.node_level.get_const_data();
+    const auto inv_permutation = inv_permutation_array.get_data();
+    const auto key = key_array.get_data();
+    const auto it = thrust::make_counting_iterator(IndexType{});
+    const auto inv_permutation_it =
+        thrust::make_permutation_iterator(inv_permutation, level_nodes);
+    thrust::fill_n(policy, inv_permutation, num_rows,
+                   std::numeric_limits<IndexType>::max());
+    // fill inverse permutation for first level
+    const auto num_components = level_ptrs[1];
+    thrust::copy_n(policy, it, num_components, inv_permutation_it);
+    for (IndexType lvl = 1; lvl < num_levels; lvl++) {
+        const auto level_begin = level_ptrs[lvl];
+        const auto level_end = level_ptrs[lvl + 1];
+        const auto level_size = level_end - level_begin;
+        // sort by node ID for determinism
+        thrust::sort(policy, level_nodes + level_begin,
+                     level_nodes + level_end);
+        // sort by degree as tie-breaker
+        thrust::copy_n(policy,
+                       thrust::make_permutation_iterator(
+                           degrees, level_nodes + level_begin),
+                       level_size, key + level_begin);
+        thrust::stable_sort_by_key(policy, key + level_begin, key + level_end,
+                                   level_nodes + level_begin);
```

Comment:
we can do this sorting step once outside of the loop (followed by a level sorting in the end) to get much fewer kernel invocations.